### PR TITLE
Property bindings and test component

### PR DESCRIPTION
**Pull request by Vlad160**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,12 +20,14 @@
   "homepage": "https://github.com/Vlad160/Simple-Components#readme",
   "dependencies": {
     "custom-elements": "^1.0.2",
-    "typescript": "^2.7.2",
-    "webpack": "^4.1.0"
+    "lodash": "^4.17.5",
+    "typescript": "^2.7.2"
```
Comment (r-karol) [inline package.json]: Can move this to devDependencies section

File: src/core/base.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+import {IChange} from "./lifecycle";
+import {toCamelCase, toKebabCase} from "./utils";
+
 export const bindings: string = '__bindings__';
 export const observedAttributes: string = 'observedAttributes';
```
Comment (r-karol) [inline src/core/base.ts]: Do we need to export these constants?

File: src/core/base.ts [hunk 3]
```diff
@@ -8,7 +11,7 @@ export function Component<T extends Function>(selector: string, options: Element
         } else {
```
Comment (r-karol) [inline src/core/base.ts]: Can we remove else block? We don't need it since error is thrown in other case

File: src/core/base.ts [hunk 4]
```diff
@@ -8,7 +11,7 @@ export function Component<T extends Function>(selector: string, options: Element
...
             target.constructor.prototype[observedAttributes] = [];
             target.prototype[bindings].forEach((prop: string) => {
```
Comment (r-karol) [inline src/core/base.ts]: Are we going to get an error if component has no binding properties?

File: src/core/base.ts [hunk 5]
```diff
@@ -34,35 +37,73 @@ export function createPropertyBinding(target: any, key: string): PropertyDescrip
...
         set(value: any) {
             this[`_${key}`] = value;
-            this.render();
+            this.renderCallback();
         }
     }
 }
 
-@Component('simple-element')
 export class SimpleComponent extends HTMLElement {
```
Comment (r-karol) [inline src/core/base.ts]: It probably should be abstract

File: src/core/base.ts [hunk 6]
```diff
@@ -34,35 +37,73 @@ export function createPropertyBinding(target: any, key: string): PropertyDescrip
...
-            this[name] = newValue;
+    protected renderCallback(): void {
+        if (this.connected) {
+            this.beforeRender();
+            this.render();
+            this.afterRender();
         }
     }
 
     protected render(): void {
```
Comment (r-karol) [inline src/core/base.ts]: Shouldn't `render`, `beforeRender` and `afterRender` be private?

File: src/core/lifecycle.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export interface IChange {
```
Comment (r-karol) [inline src/core/lifecycle.ts]: Maybe lets rename it to `AttributeValueChange`? (with no prefix of 'I', saw it in Angular Material and kinda like it)

File: src/core/base.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        if (!target[bindings]) {
+            target[bindings] = [];
+        }
+        if (!target[bindings][key]) {
+            target[bindings].push(key);
+        }
+    }
+}
+
+export function createPropertyBinding(target: any, key: string): PropertyDescriptor {
```
Comment (r-karol) [inline src/core/base.ts:33]: Do we need to export it?

File: package.json (hunk 1 above)
Comment (Vlad160) [reply package.json]: Done

File: src/core/base.ts (hunk 2 above)
Comment (Vlad160) [reply src/core/base.ts]: Done

File: src/core/base.ts (hunk 3 above)
Comment (Vlad160) [reply src/core/base.ts]: No. It's a basic check.

File: src/core/base.ts (hunk 4 above)
Comment (Vlad160) [reply src/core/base.ts]: Fixed

File: src/core/base.ts (hunk 5 above)
Comment (Vlad160) [reply src/core/base.ts]: Sure. Done

File: src/core/base.ts (hunk 6 above)
Comment (Vlad160) [reply src/core/base.ts]: No. It's part of lifecycle

File: src/core/lifecycle.ts (hunk 7 above)
Comment (Vlad160) [reply src/core/lifecycle.ts]: Okay

**Review by r-karol — APPROVED** [review]
Great job! Well done!

**Pull request merged**